src/Graph/Trees/Heap.php

```diff
@@ -0,0 +1,179 @@
+     *
+     * @return mixed[]|null
+     */
+    public function heappop() : ?array
+    {
+        if (!is_array($this->heap) || empty($this->heap)) {
+            return null;
+        }
+
+        $last_element = array_pop($this->heap);
+
+        if (!is_array($this->heap) || empty($this->heap)) {
+            return $last_element;
+        }
+
+        $return_item = $this->heap[0];
+
+        $this->heap[0] = $last_element;
+        $this->heap = $this->siftup($this->heap, 0);
+
+        return $return_item;
+    }
+
+    /**
+     * Pushes the value of the item in the heap.
+     *
+     * @param mixed[] $item
+     * @return mixed[]
+     */
+    public function heappush(array $item) : array
+    {
+        $this->heap[] = $item;
+        $new_count = (count($this->heap) - 1);
+
+        $this->siftdown(0, $new_count);
+
+        return $this->heap;
+    }
+
+    /**
+     * Finds the best fit for the new item & updates heap while maintaining heap invariant.
+     *
+     * @param int $start_pos
+     * @param int $i
+     */
+    private function siftdown(int $start_pos, int $i) : void
+    {
+        $temp_item = $this->heap[$i];
+
+        while ($i > $start_pos) {
+            $parent_pos = ($i - 1) >> 1;
+            $parent = $this->heap[$parent_pos];
+
+            if (($temp_item['count'] < $parent['count'])) {
+                $this->heap[$i] = $parent;
+                $i = $parent_pos;
+
+                continue;
+            }
+
+            break;
+        }
+
+        $this->heap[$i] = $temp_item;
+    }
+
+    /**
+     * Creates a heap at the provided index while maintaining heap invariant.
+     *
+     * @param mixed[] $x
+     * @param int $i
+     * @return mixed[]
+     */
+    private function siftup(array $x, int $i) : array
+    {
+        $end_pos = count($x);
+        $start_pos = $i;
+        $new_item = $this->heap[$i];
+        $child_pos = (2 * $i) + 1;
+
+        while ($child_pos < $end_pos) {
+            $right_pos = $child_pos + 1;
+
+            if (empty($this->heap[$right_pos]['count'])) {
+                $right_heap = 0;
+            } else {
+                $right_heap = $this->heap[$right_pos]['count'];
+            }
+            
+            if ($right_pos < $end_pos && ($this->heap[$child_pos]['count'] >= $right_heap)) {
+                $child_pos = $right_pos;
+            }
+
+            $this->heap[$i] = $this->heap[$child_pos];
+            $i = $child_pos;
+            $child_pos = (2 * $i) + 1;
+        }
+
+        $this->heap[$i] = $new_item;
+        $this->siftdown($start_pos, $i);
+
+        return $this->heap;
+    }
+
+    /**
+     * Converts an array into a heap.
+     *
+     * @param array[] $x
+     * @return mixed[]
+     */
+    private function heapify(array $x) : array
+    {
+        $this->heap = array_values($x);
+        $n = count($this->heap);
+
+        $max_range = (($n / 2) - 1);
+
+        foreach (array_reverse(range(0, $max_range)) as $i) {
```

Comment:
It would be but hopping back to your previous statements regarding making this efficient after launching it, since this is a one-time step that's not involved in the training loop, the incremental time isn't too big of a deal.

I can update however if you'd like.

Reply:
Does the '--$i' reverse the $i variable?

Reply:
`--$i` pre decrements the $i variable ... in the case of that for loop $i is counting backward by 1 from $maxRange

You could even do ..

```php
foreach (range($max_range, 0) as $i) {
```

and it would avoid the array_reverse (an unnecessary O(n) operation), however the for loop is optimal in memory as well as time so let's use that unless you can think of a good reason not to.